### PR TITLE
Context threaded

### PR DESCRIPTION
**Pull request by shannynalayna**
Current progress on the threaded Context Manager.

Curious about (and hoping for feedback/thoughts on) the following items:

- Organization of Context Manager vs the Context Manager thread, especially where the function to drop / shutdown the Context Manager thread should be implemented.
- Naming (re: SchedulerAgent, ContextExecution trait, ExecuterAgent)
- Errors, Should the ContextManagerErrors (following the ContextLifecycleError) get broken down further into the ContextExecutionErrors and ContextManagerThreadError?

### CODE REVIEW

File: libtransact/src/context/executer_agent.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    pub fn new(
+        tx: Sender<ContextOperationMessage>,
+        rx: Receiver<ContextOperationResponse>,
+    ) -> Self {
+        ExecuterAgent { tx, rx }
+    }
+}
+
+impl ContextExecution for ExecuterAgent {
+    fn get(
```
Comment (vaporos) [inline libtransact/src/context/executer_agent.rs]: Shoudl this be get_state() to match the other functions?

File: libtransact/src/context/manager.rs [hunk 2]
```diff
@@ -222,52 +193,59 @@ impl ContextManager {
     }
 }
 
-pub mod sync {
-    //! This module provides a thread-safe ContextManager.
-    //!
-    //! For many uses of the context manager, it will need to be shared between multiple threads,
-    //! with some threads reading and writing to a context while others create contexts.
+pub mod lifecycle {
```
Comment (vaporos) [inline libtransact/src/context/manager.rs]: I think we should keep modules in their own files. Which would give us context/manager/mod.rs, context/manager/lifecycle.rs, context/manager/sync.rs, context/manager/thread.rs.

We could move things around in a separate PR after this one.

File: libtransact/src/context/mod.rs [hunk 3]
```diff
@@ -20,25 +20,74 @@
 pub type ContextId = [u8; 16];
 
 mod error;
+mod executer_agent;
 pub mod manager;
+mod scheduler_agent;
```
Comment (vaporos) [inline libtransact/src/context/mod.rs]: I'm not a fan of the scheduler_agent/executer_agent naming as it names these modules in terms of the external modules they may interact with -- they could be used by other things in the future, so it would be better to use context-centric terms.

We could rename these later in a different PR if we think of better names.

File: libtransact/src/scheduler/serial/core.rs [hunk 4]
```diff
@@ -17,7 +17,7 @@
 
 //! Implementation of core scheduler thread.
 
-use crate::context::manager::ContextManagerError;
+use crate::context::manager::ContextLifecycleError;
```
Comment (vaporos) [inline libtransact/src/scheduler/serial/core.rs]: This is a good example of the 'manager' piece of the API leaking into the scheduler implementation. Per suggestion above, this would turn into context::lifecycle::ContextLifecycleError. But if we keep ContextLifecycle in the top-level module, we should just be able to do 'use crate::context::{ContextId, ContextLifecycle, ContextLifecycleError};' here.

File: libtransact/src/context/manager.rs (hunk 2 above)
Comment (shannynalayna) [reply libtransact/src/context/manager.rs]: I can separate the sync module into its own file in a separate pr and then move the lifecycle module into another file for this pr.

File: libtransact/src/context/executer_agent.rs (hunk 1 above)
Comment (shannynalayna) [reply libtransact/src/context/executer_agent.rs]: Do you think this should be changed from the Context Manager up, or just for this external API?

File: libtransact/src/context/manager/thread/mod.rs [hunk 5]
```diff
@@ -295,3 +295,169 @@ impl ContextManagerThread {
...
+    static EVENT_TYPE1: &str = "sawtooth/block-commit";
+    static ATTR1: (&str, &str) = (
+        "block_id",
+        "f40b90d06b4a9074af2ab09e0187223da7466be75ec0f472 \
+         f2edd5f22960d76e402e6c07c90b7816374891d698310dd25d9b88dce7dbcba8219d9f7c9cae1861",
+    );
+    static ATTR2: (&str, &str) = ("block_num", "3");
+
+    fn shutdown_thread(join_handle: JoinHandle<()>) {
+        thread::Builder::new()
```
Comment (peterschwarz) [inline libtransact/src/context/manager/thread/mod.rs]: This should not spawn a new thread, but just join the thread directly.

It also doesn't actually signal for the context manager thread to shut down, so if you remove it, the test hangs.

**Comment by peterschwarz** [comment]
We should also look at what should be publicly accessible.  Do those `*_agent` modules need to be public?

File: libtransact/src/context/manager/thread/mod.rs (hunk 5 above)
Comment (shannynalayna) [reply libtransact/src/context/manager/thread/mod.rs]: The tests hang if this function just joins directly, hence spawning a new thread. This pertains to one of the questions I had for this PR, in how to handle shutting down this thread. Wasn't quite sure what best practice would be in this case.

Comment (peterschwarz) [reply libtransact/src/context/manager/thread/mod.rs]: It should signal the background thread using the `Shutdown` enum value
